Chat visual update

Banana090 opened the pull request:
**Includes some critical bugs in it :smirk: :smirk:**

- Player's skin near message author's name
- Transparent black background for chat
- Text margin when message is multiline
- Messages margin

![Screenshot_2](https://user-images.githubusercontent.com/39315809/95912054-f4f17b80-0daa-11eb-937b-48828c8a4d18.png)

Banana090 commented:
Already found and fixed infinite loop bug when drawing too long message with no spaces

heinrich5991 commented:
I *think* I like these graphics for chat, I'm not sure whether they'll be too distracting though.

Maybe the border could only be shown while the chat input is open?

Only throwing ideas out here, don't start implementing them right away unless you also think they're good ideas.

What do others think about this graphics change? @Learath2 @Jupeyy @edg-l (random pings)

heinrich5991 commented:
Thanks for your first contribution btw! It's quite a big change for your first contribution. ;)

Banana090 commented:
![Screenshot_4](https://user-images.githubusercontent.com/39315809/95927244-e31dd180-0dc6-11eb-91ab-323142682614.png)

Sorry for bad screenshot quality, but I don't have another for now. For animated demo see discord [showroom]: https://discordapp.com/channels/252358080522747904/295908390956433410/765719953188978748

Fixed bugs, made backgrounds for each message individually. For me it looks nice and don't distract from game.

Commit does still have 2 bugs which I know about and will fix asap.
1. Wrong message width calculation in some cases
2. When there is a message from player still in buffer, but player leaves, you'll have white texture instead of his skin

heinrich5991 commented:
Direct link to the video: https://cdn.discordapp.com/attachments/295908390956433410/765719948403671040/2020-10-14_02-26-49.mp4.

Jupeyy commented on src/game/client/components/chat.cpp:
```diff
@@ -848,13 +849,22 @@ void CChat::OnPrepareLines()
-				str_format(aName, sizeof(aName), "%d: ", m_aLines[r].m_ClientID);
-			else
-				str_format(aName, sizeof(aName), " %d: ", m_aLines[r].m_ClientID);
+			m_aLines[r].m_AuthorRenderInfo = m_pClient->m_aClients[m_aLines[r].m_ClientID].m_RenderInfo;
```
This is not safe anymore, renderinfo is allowed to change, e.g. by reloading skins, see gameclient "RefindSkins"

Banana090 replied:
I need some help with that and more info. Can you provide an example of how should I render tee with needed size? Or contact me on the DDNet Discord

Jupeyy commented on src/engine/client/text.cpp:
```diff
@@ -1299,7 +1306,7 @@ class CTextRender : public IEngineTextRender
 					++CharacterCounter;
 
-					DrawX = pCursor->m_StartX;
+					DrawX = pCursor->m_StartX + +pCursor->m_NewLineOffsetX;
```
I think, you can just appendtext to the text container with a new cusor, it will start at the StartX of that new cursor

Banana090 replied:
Could be ok, but it will require one more draw call for that, as well as creating new cursor and second container. Sounds like unnecessary actions

Jupeyy replied:
it would require another draw call, it appends the text the the existing text container, the extra cursor is probably more efficient than adding code to a function that is called by other code too :D

Banana090 replied:
it's called by other functions with 0 offset, so takes no effect at all

Banana090 replied:
But maybe you're right, yeah

Jupeyy replied:
My point is also more about not bloating these functions too much, on the good side:
I really like when the text is aligned properly after the tee name, looks much cleaner

Banana090 replied:
It's not so easy btw, because AppendTextContainer function can break lines by itself, so it moves the cursor. If we want to use new cursor for 2nd+ line of message then we'll need to copy all break-line code into chat.cpp which is stupid. So I think it's better to stay with my current solution on this @Jupeyy

Jupeyy commented on src/game/client/components/chat.cpp:
```diff
@@ -1069,6 +1085,37 @@ void CChat::OnRender()
+	float LineWidth = m_pClient->m_pScoreboard->Active() ? 90.0f : 200.0f;
+	float BackgroundY = 300 - 28;
+
+	for(int i = 0; i < MAX_LINES; i++)
```
not sure about this second loop, can't it be merged with the other loop?

Banana090 replied:
Did you check very first commit? Second loop was already deleted and then added again, lol) Now it is used for batching of backgrounds. Should run some performance test probably to see if it increases performance and if it doesn't then yes, it could be merged easily

Jupeyy commented on src/game/client/components/chat.cpp:
```diff
@@ -1085,6 +1132,9 @@ void CChat::OnRender()
 
 		if(m_aLines[r].m_TextContainerIndex != -1)
 		{
+			if(m_aLines[r].m_NeedRenderTee)
```
Defenitelly make this optional, and maybe we should even discuss to make it off default.
The community is very sensitive to such changes.

Jupeyy commented on src/game/client/components/chat.cpp:
```diff
@@ -1069,6 +1085,37 @@ void CChat::OnRender()
+	}
+
+	if(Now <= m_aLines[m_CurrentLine % MAX_LINES].m_Time + 16 * time_freq() || m_PrevShowChat)
+	{
```
Like heinrich said, maybe only show the background rect, when we are activly chatting?

Banana090 replied:
Disagreed. And again, looks like you're looking on the first commit.
Now there is a more transparent background for every message that uses message's width. I think it looks nice and don't distract at all. But ppl need to see it by themselves

Jupeyy replied:
it doesnt matter where the background is

Banana090 replied:
I thought about making an option for that, so I'll need to have special tab in setting just for chat. The problem is message marging. If you'll disable background rendering then messages will look like shit I suppose

Banana090 replied:
![Screenshot_5](https://user-images.githubusercontent.com/39315809/95974313-800b5980-0e1d-11eb-8765-4be3f5d1ba5c.png)
Or it's not. Not sure, but anyway it's differs from what is was before

Jupeyy reviewed (changes requested):
I am generally not against this change, but i cannot tell how sensetive ppl will react to this

edg-l commented:
You should definitely test how it looks when:
- Show client ids on/off
- Friend heart
- Repetition counter

(and their combinations)

The image where text have their own rounded rect (each line) and is not uniform maybe looks better if each line has a margin (the background rects dont touch), or maybe simply make the rounded rect not independent.

Banana090 commented:
> You should definitely test how it looks when:
> 
> * Show client ids on/off
> * Friend heart
> * Repetition counter
> 
> (and their combinations)
> 
> The image where text have their own rounded rect (each line) and is not uniform maybe looks better if each line has a margin (the background rects dont touch), or maybe simply make the rounded rect not independent.

@edg-l 

Of course I'm testing it. Friend heart have some problems now, because I simply can't imagine a good design solution where to replace it or maybe even show another way.

show ids is fine, as well as repetition counter

Tee icon also have some problems now, but when it's fixed it should not update the skin for past messages I think.

Banana090 commented:
As previous commit says, I've added 3 config variables

- cl_chat_tees 0 1
- cl_chat_backgrounds 0 1
- cl_chat_margin 0 1

So now you can use any combination of new features/settings or disable them completely

Learath2 commented:
I think this looks nice enough. I don't mean to discourage you but getting this kind of change in creates a looot of friction, I think  you'll have to add a way to disable every single new thing before this is palatable to the oldguard that hates every change, especially visual ones.

def- commented:
We could make a vote in #announcements to see if this is wanted. But probably we should have done that before the implementation was done.

heinrich5991 commented:
@Learath2 I disagree with providing configuration options for every single possibility. It will make change here impossible.

I'd vote for either (1) no new option, or (2) reconciling old options to only allow selection of how the default options looked at any time, so 1: very old, plain chat, 2: plain chat + hearts, 3: plain chat + hearts + message repetition counter, 4: this new chat. This could at least keep the number of configuration options low.

Banana090 commented:
@heinrich5991 what if you like the new chat but don't want tee icons in messages?

And why do you even need setting to fallback to plain text or disabling repeat counter

heinrich5991 commented:
@Banana090 Sorry, I'm not asking you to implement this config option, I'm just wondering how to best solve our problem of exploding amounts of config options.

The problem with introducing more and more config options to selectively disable certain features is that it becomes hard to impossible to change the existing code without breaking something, hindering the exploration of new features such as your PR. I'm trying to solve that problem by offering less configuration options, but still allowing players to go back to a version they liked.

> And why do you even need setting to fallback to plain text or disabling repeat counter?
> I mean what's the reason for making type for text + hearts if it's already cl_friend_messages

I'm proposing to remove these config options so we can reduce the amount of options we have to consider when changing that code. Again: I'm not asking you to implement this new config option!

heinrich5991 commented:
@Learath2 @def- What would you think about my idea for reducing the amount of config options?

Banana090 commented:
@heinrich5991 yes, that's what I'm talking about. Figuring out solutions for reducing config options but still providing a way to play with older styles.
What if you don't like old version and like new one, but don't want to see tee icons? Just deal with it and get used to?

heinrich5991 commented:
Yes, basically. In order to keep the maintenance level sane.

Banana090 commented:
Sounds logical enough for me, I don't mind implementing this

heinrich5991 commented:
Please wait with an implementation until we reach a consensus whether we actually want to do this.

def- commented:
@heinrich5991 there will definitely be people who want feature x and y, but not z, which will be impossible for them. I don't see a good solution honestly, except only doing changes that are strictly better. (but that's nearly impossible)

heinrich5991 commented:
@def- My idea is that this way, changes are always improvements or keeping the status quo. No one gets a worse interface, and the whole thing stays a little bit more maintainable.

Banana090 commented:
- Fixed everything
- Cleaned up the code
- Optimized (even old code)
- Added cl_chat_old config variable for old chat style
- Removed all my previous config variables

This is my pre-last commit. The only thing left is to add actual toggle in Settings > Hud
After adding that I think I will not work with the chat branch anymore. At least for some time. Will be moving on new things.

You can take some time and discuss this PR, even decline that, but I see no reason for not merging it in near future. Maybe this would need just some code tweaks that I've done wrongly due to my knowledge of C++

About general discussion of making visual updates - it already has a setting to fallback for the old style. I see enough people looking forward to play with new chat style, so it's definitely good update and if someone don't like it they can just turn it off and be happy <3

Banana090 commented:
There is still a bug I found, when holding [tab] with new chat style there is a chance that one of shown messages will not align properly on Y axis. Did not find any way to fix. Somehow affected by line 1033 of chat.cpp

```cpp
				TextRender()->AppendTextContainer(&Cursor, m_aLines[r].m_TextContainerIndex, aCount);
		}

		if(!g_Config.m_ClChatOld)
>>>			Cursor.m_StartX = Cursor.m_X;

		// render line
		ColorRGBA Color;
```

Jupeyy commented on src/engine/client/text.cpp:
```diff
@@ -1401,6 +1417,9 @@ class CTextRender : public IEngineTextRender
 					pCursor->m_GlyphCount++;
 					++CharacterCounter;
 				}
+
```
so the idea is, that this is some kind of LongestLine ?, cant it be done in the cursor instead of the text container?

Jupeyy commented on src/game/client/components/chat.cpp:
```diff
@@ -1069,6 +1156,44 @@ void CChat::OnRender()
+			{
+				STextContainerSize LineSize = TextRender()->GetTextContainerSize(m_aLines[r].m_TextContainerIndex);
+				Graphics()->SetColor(0, 0, 0, 0.12f * Blend);
+				RenderTools()->DrawRoundRect(4.5f, bgY - Padding / 2.0f, LineSize.m_Width, LineSize.m_Height + Padding, Rounding);
```
i'd still refer a quadcontainer here, but if thats too confusing, i'll clean it up later

Jupeyy reviewed (commented):
I'll test the functionallity later, but i guess its pretty good already

def- commented:
I added a vote in #announcements on Discord.

def- commented:
![screenshot-20201015@181832](https://user-images.githubusercontent.com/2335377/96157671-e02ff780-0f12-11eb-8456-763c1b1b98c8.png)
Seems like the length calculation is a bit off, 3 lines but only enough space for 2.

def- commented:
Tee graphics bug when you open chat, maybe the person changed skin?
![screenshot-20201015@182641](https://user-images.githubusercontent.com/2335377/96158728-17eb6f00-0f14-11eb-904c-f282a539ec20.png)

Banana090 commented:
it worked for me on skin change, not sure. Will make some tests

ChillerDragon commented:
Looks good on the screenshot. I am sure the skins can be helpful to distinguish between similar named tees sometimes :)

good job @Banana090
